Add coordinate transform support to general field

GiudGiud opened the pull request:
seems to work so far except UOs
EDIT: UOs fixed, original transfers had a bug, not new ones

GiudGiud committed:
Add support for coordinate collapsing to transfers
- source transformation in general field

GiudGiud committed:
Add support for nearest-location general field transfers
refs #12770

GiudGiud committed:
Add tests for coordinate transformations with general field nearest location
refs #12770

GiudGiud committed:
Add support for coord transform in hape evaluation generation field transfers
 refs #12770

GiudGiud committed:
Port old test for shape evaluation coord transform

GiudGiud committed:
Add support for coord transform in general field user object transfers
     refs #12770

GiudGiud committed:
Port old test for user object transfer & coord transform

GiudGiud committed:
Add utility routine to get the global ids to get the coord transforms

GiudGiud committed:
Fixup coord transforms

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1859309/) on be1b137 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25945/site)

This comment will be updated on new commits.

GiudGiud committed:
Remove bug on the user object transfer coordinate transform
Coordinate transform was being applied twice

GiudGiud committed:
Add more comments to the general field transfer code
Add more information in the nearest_position cases
Add more information in search value conflict logs

lindsayad commented on framework/include/transfers/MultiAppTransfer.h:
```diff
@@ -188,6 +192,17 @@ class MultiAppTransfer : public Transfer
   std::vector<unsigned int> _from_local2global_map;
 
+  /// Return the global app index from the local index in the "from-multiapp" transfer direction
+  unsigned int getGlobalSourceAppIndex(unsigned int i_from) const
+  {
+    return _current_direction == TO_MULTIAPP ? 0 : _from_local2global_map[i_from];
+  }
+  /// Return the global app index from the local index in the "to-multiapp" transfer direction
+  unsigned int getGlobalTargetAppIndex(unsigned int i_to) const
+  {
+    return _current_direction == FROM_MULTIAPP ? 0 : _to_local2global_map[i_to];
```
should it error if the boolean is `true`?

lindsayad replied:
never mind

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.C:
```diff
@@ -136,6 +138,10 @@ MultiAppGeneralFieldShapeEvaluationTransfer::evaluateInterpValuesWithMeshFunctio
+          if (!_skip_coordinate_collapsing &&
+              _from_transforms[from_global_num]->hasNonTranslationTransformation())
+            mooseError(
+                "Rotation and scaling currently unsupported with nearest positions transfer.");
```
This is a `ShapeEvaluationTransfer`

lindsayad replied:
but maybe this message is still correct

GiudGiud replied:
yup. nearest-positions can still be done for shapeEvaluation + overlapping domains. 
Howver it s kind of a stretch

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldShapeEvaluationTransfer.C:
```diff
@@ -125,7 +125,9 @@ MultiAppGeneralFieldShapeEvaluationTransfer::evaluateInterpValuesWithMeshFunctio
       {
         // Use mesh function to compute interpolation values
-        auto val = (local_meshfuns[i_from])(pt - _from_positions[i_from]);
+        const auto from_global_num =
+            _current_direction == TO_MULTIAPP ? 0 : _from_local2global_map[i_from];
```
Are you going to use the new methods you made?

GiudGiud replied:
I added them after! I ll do it now

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -688,7 +696,18 @@ MultiAppGeneralFieldTransfer::cacheIncomingInterpVals(
-          const auto p = point_requests[val_offset] - _to_positions[problem_id];
+          const auto & to_transform = _to_transforms[getGlobalTargetAppIndex(problem_id)];
+          Point p;
+          if (to_transform->hasNonTranslationTransformation())
+          {
+            if (!_skip_coordinate_collapsing)
+              mooseInfo("Search value conflict cannot find the origin point due to the "
+                        "non-uniqueness of the coordinate collapsing reverse mapping");
+            p = point_requests[val_offset] - _to_positions[problem_id];
```
I don't understand this. We don't use the transform if there is non-translation information? But we do if it doesn't?

GiudGiud replied:
changed. it s really the coordinate collapse back I dont want to do, because it errors on RZ -> XYZ but I do want to be doing that direction for now even if it s ill posed. maybe I shouldnt..

GiudGiud replied:
it s for the search value conflict detection, it s not for the main values

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -848,8 +873,18 @@ MultiAppGeneralFieldTransfer::examineLocalValueConflicts(
-      auto local_p = p + _to_positions[i_to];
+      // Move to the local frame of reference for the target
+      Point local_p;
+      const auto & to_transform = _to_transforms[getGlobalTargetAppIndex(i_to)];
+      if (to_transform->hasNonTranslationTransformation())
+      {
+        if (!_skip_coordinate_collapsing)
+          mooseInfo("Resolution of local value conflicts detected cannot find the exact origin "
+                    "point due to the non-uniqueness of the coordinate collapsing reverse mapping");
+        local_p = p - _to_positions[i_to];
+      }
+      else
+        local_p = to_transform->mapBack(p);
```
still don't get it haha. Also if this is going to be a repeated pattern, is it possible to create a method and use it instead of repeating?

GiudGiud replied:
added new APIs

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -1168,23 +1219,23 @@ MultiAppGeneralFieldTransfer::acceptPointInOriginMesh(unsigned int i_from,
   else
   {
     auto * pl = _from_point_locators[i_from].get();
+    const auto from_global_num =
+        _current_direction == TO_MULTIAPP ? 0 : _from_local2global_map[i_from];
```
use your method

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -1415,9 +1468,11 @@ MultiAppGeneralFieldTransfer::getRestrictedFromBoundingBoxes()
-      bbox.second += _from_positions[j];
+      // Translate the bounding box to the from domain's position. We may have rotations so we
+      // must be careful in constructing the new min and max (first and second)
+      const auto from_global_num =
+          _current_direction == TO_MULTIAPP ? 0 : _from_local2global_map[j];
```
your method

lindsayad commented on framework/src/transfers/MultiAppUserObjectTransfer.C:
```diff
@@ -528,7 +528,7 @@ MultiAppUserObjectTransfer::execute()
             {
               Moose::ScopedCommSwapper swapper(getFromMultiApp()->comm());
-              from_value = user_object.spatialValue(from_transform.mapBack(to_transform(point)));
+              from_value = user_object.spatialValue(from_transform.mapBack(point));
```
was it already transformed?

lindsayad reviewed (changes requested):
what motivated you to incorporate some of my least-proud-of-code?

GiudGiud commented:
> what motivated you to incorporate some of my least-proud-of-code?

Stefano wants it

GiudGiud committed:
Address Alex's review
- use new api to get global index
- factor out target app point search
- add an API to look at the coordinate change

lindsayad commented on framework/src/transfers/MultiAppTransfer.C:
```diff
@@ -609,3 +609,24 @@ MultiAppTransfer::checkVariable(const FEProblemBase & fe_problem,
+  if (to_transform->hasCoordinateSystemTypeChange())
+  {
+    if (!_skip_coordinate_collapsing)
+      mooseInfo(phase + " cannot use the point in the target app frame to the "
```
"app frame to the" -> "app frame due to the"

GiudGiud replied:
changed

lindsayad reviewed (commented):
This is much clearer to me now, thank you!

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1861210/) on 515aa30 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>191b9d</th>
      <th colspan=4>#25945 515aa3</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.68%</td>
      <td>+0.04%</td>
      <td>93.02%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92451</td>
      <td>92532</td>
      <td>+81</td>
      <td>80</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15509</td>
      <td>15471</td>
      <td>-38</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25945/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25945/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25945/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
Thanks for all the reviews this afternoon!
